test/e2e/framework/framework.go

```diff
@@ -0,0 +1,134 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package framework is a package helping setting up end-to-end testing accross a
+// Kubernetes cluster.
+package framework
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"agones.dev/agones/pkg/apis/stable/v1alpha1"
+	"agones.dev/agones/pkg/client/clientset/versioned"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	// required to use gcloud login see: https://github.com/kubernetes/client-go/issues/242
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// Framework is a testing framework
+type Framework struct {
+	KubeClient      kubernetes.Interface
+	AgonesClient    versioned.Interface
+	GameServerImage string
+}
+
+// New setups a testing framework using a kubeconfig path and the game server image to use for testing.
+func New(kubeconfig, gsimage string) (*Framework, error) {
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "build config from flags failed")
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating new kube-client failed")
+	}
+
+	agonesClient, err := versioned.NewForConfig(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating new agones-client failed")
+	}
+
+	return &Framework{
+		KubeClient:      kubeClient,
+		AgonesClient:    agonesClient,
+		GameServerImage: gsimage,
+	}, nil
+}
+
+// CreateGameServerAndWaitUntilReady Creates a GameServer and wait for its state to become ready.
+func (f *Framework) CreateGameServerAndWaitUntilReady(ns string, gs *v1alpha1.GameServer) (*v1alpha1.GameServer, error) {
+	newGs, err := f.AgonesClient.StableV1alpha1().GameServers(ns).Create(gs)
+	if err != nil {
+		return nil, fmt.Errorf("creating %v GameServer instances failed (%v): %v", gs.Spec, gs.Name, err)
+	}
+
+	readyGs, err := f.WaitForGameServerState(newGs, v1alpha1.Ready, 5*time.Minute)
+
+	if err != nil {
+		return nil, fmt.Errorf("waiting for %v GameServer instance readiness timed out (%v): %v",
+			gs.Spec, gs.Name, err)
+	}
+
+	return readyGs, nil
+}
+
+// WaitForGameServerState Waits untils the gameserver reach a given state before the timeout expires
+func (f *Framework) WaitForGameServerState(gs *v1alpha1.GameServer, state v1alpha1.State,
+	timeout time.Duration) (*v1alpha1.GameServer, error) {
+	var pollErr error
+	var readyGs *v1alpha1.GameServer
+
+	err := wait.Poll(2*time.Second, timeout, func() (bool, error) {
```

Comment:
Should we `Poll` or `PollImmediate` ? If we `Poll` we always wait for 2 second, when maybe we don't need to?

Reply:
yes you're right ! let me change that